test/integration/ThreeChainEndToEnd.t.sol

```diff
@@ -0,0 +1,213 @@
+
+        vm.startPrank(BSM);
+        sB.gateway.subsidizePool{value: DEFAULT_SUBSIDY}(POOL_A);
+        sC.gateway.subsidizePool{value: DEFAULT_SUBSIDY}(POOL_A);
+    }
+
+    /// @notice Test transferring shares between Chain B and Chain C via Hub A
+    /// forge-config: default.isolate = true
+    function testTransferSharesWithHop() public {
+        uint128 amount = 1000 * 1e18;
+
+        testConfigurePoolWithThreeChains();
+
+        // B + C: Deploy vaults on both chains
+        vm.startPrank(FM);
+        AssetId USDC_ID_B = newAssetId(sB.centrifugeId, 1);
+        h.hub.updateVault{value: GAS}(POOL_A, SC_1, USDC_ID_B, sB.asyncVaultFactory, VaultUpdateKind.DeployAndLink);
+        AssetId USDC_ID_C = newAssetId(sC.centrifugeId, 1);
+        h.hub.updateVault{value: GAS}(POOL_A, SC_1, USDC_ID_C, sC.asyncVaultFactory, VaultUpdateKind.DeployAndLink);
+        vm.stopPrank();
+
+        // B: Mint shares
+        vm.startPrank(address(sB.root));
+        IShareToken shareTokenB = IShareToken(sB.spoke.shareToken(POOL_A, SC_1));
+        shareTokenB.mint(INVESTOR_A, amount);
+        vm.stopPrank();
+        assertEq(shareTokenB.balanceOf(INVESTOR_A), amount, "Investor should have minted shares on chain B");
+
+        // B: Initiate transfer of shares
+        vm.expectEmit();
+        emit ISpoke.TransferShares(sC.centrifugeId, POOL_A, SC_1, INVESTOR_A, INVESTOR_A.toBytes32(), amount);
+        emit IHub.ForwardTransferShares(sC.centrifugeId, POOL_A, SC_1, INVESTOR_A.toBytes32(), amount);
+        vm.expectEmit(true, false, false, false);
+        emit IGateway.UnderpaidBatch(sC.centrifugeId, bytes(""));
+        vm.prank(INVESTOR_A);
+        sB.spoke.transferShares{value: GAS}(sC.centrifugeId, POOL_A, SC_1, INVESTOR_A.toBytes32(), amount);
+
+        assertEq(shareTokenB.balanceOf(INVESTOR_A), 0, "Shares should be burned on chain B");
+
+        // C: Transfer expected to be pending on A due to message being unpaid
+        IShareToken shareTokenC = IShareToken(sC.spoke.shareToken(POOL_A, SC_1));
+        assertEq(shareTokenC.balanceOf(INVESTOR_A), 0, "Share transfer not executed due to unpaid message");
+
+        // A: Before calling repay, set a refund address for the pool
+        vm.prank(address(h.root));
+        h.gateway.setRefundAddress(POOL_A, IRecoverable(h.gateway));
+
+        // A: Repay for unpaid ExecuteTransferShares message on A to trigger sending it to C
+        bytes memory message = MessageLib.ExecuteTransferShares({
+            poolId: PoolId.unwrap(POOL_A),
+            scId: ShareClassId.unwrap(SC_1),
+            receiver: INVESTOR_A.toBytes32(),
+            amount: amount
+        }).serialize();
+        vm.expectEmit(true, false, false, false);
+        emit IMultiAdapter.HandlePayload(h.centrifugeId, bytes32(""), bytes(""), adapterC);
+        vm.expectEmit();
+        emit IERC20.Transfer(address(0), INVESTOR_A, amount);
+        emit ISpoke.ExecuteTransferShares(POOL_A, SC_1, INVESTOR_A, amount);
```

Comment:
From EndToEnd perspective, I think we do not need to check these events. As long as the amount reaches the destination, it works. Dispatching the event is already tested in the UTs

Reply:
> At this point in time (i.e. shortly before launch), I see e2e tests as an assurance of a consistent full workflow

Yeah, I also understand you. I have a very theoretical PoV regarding tests.

> we won't ever have unit tests covering all the three associated events IMO because that would be OOS for unit tests

Yeah, that's other point to work for me. I feel we move a lot of things from UT to IT.

> At this point however, we can assume minimal changes to the existing codebase and therefore I strongly argue for meaningful e2e tests that go beyond just asserting start and end amounts.

The same we said about liquidity-pools 😛 . Kidding. I understand your point. No strong opinion. Let's see how everything evolves!